src/modules/fw_auto_trim/FwAutoTrim.cpp

```diff
@@ -0,0 +1,308 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "FwAutoTrim.hpp"
+
+#include <mathlib/mathlib.h>
+
+using namespace time_literals;
+using matrix::Vector3f;
+
+FwAutoTrim::FwAutoTrim(bool is_vtol) :
+	ModuleParams(nullptr),
+	WorkItem(MODULE_NAME, px4::wq_configurations::nav_and_controllers),
+	_vehicle_torque_setpoint_sub(this, ORB_ID(vehicle_torque_setpoint), is_vtol ? 1 : 0)
+{
+	updateParams();
+	reset();
+}
+
+FwAutoTrim::~FwAutoTrim()
+{
+	perf_free(_cycle_perf);
+}
+
+bool FwAutoTrim::init()
+{
+	if (!_vehicle_torque_setpoint_sub.registerCallback()) {
+		PX4_ERR("callback registration failed");
+		return false;
+	}
+
+	_auto_trim_status_pub.advertise();
+
+	return true;
+}
+
+void FwAutoTrim::reset()
+{
+}
+
+void FwAutoTrim::updateParams()
+{
+	ModuleParams::updateParams();
+}
+
+void FwAutoTrim::Run()
+{
+	if (should_exit()) {
+		_vehicle_torque_setpoint_sub.unregisterCallback();
+		exit_and_cleanup();
+		return;
+	}
+
+	if (_vehicle_land_detected_sub.updated()) {
+		vehicle_land_detected_s vehicle_land_detected;
+
+		if (_vehicle_land_detected_sub.copy(&vehicle_land_detected)) {
+			_landed = vehicle_land_detected.landed;
+		}
+	}
+
+
+	vehicle_torque_setpoint_s vehicle_torque_setpoint;
+
+	if (!_vehicle_torque_setpoint_sub.update(&vehicle_torque_setpoint)
+	    || (vehicle_torque_setpoint.timestamp == _timestamp_last)) {
+		return;
+	}
+
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+		updateParams();
+	}
+
+	perf_begin(_cycle_perf);
+
+	if (_vehicle_status_sub.updated()) {
+		vehicle_status_s vehicle_status;
+
+		if (_vehicle_status_sub.copy(&vehicle_status)) {
+			_armed = (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
+			_fixed_wing = (vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING);
+			_tailsitter = vehicle_status.is_vtol_tailsitter;
+		}
+	}
+
+	if (_airspeed_validated_sub.updated()) {
+		airspeed_validated_s airspeed_validated;
+
+		if (_airspeed_validated_sub.copy(&airspeed_validated)) {
+			_airspeed = airspeed_validated.true_airspeed_m_s;
```

Comment:
Don't we want to use calibrated airspeed? The min/max are in CAS.

Reply:
good point